src/profile-logic/profile-data.js

```diff
@@ -685,18 +685,25 @@ export function getSampleIndexClosestToTime(
   samples: SamplesTable,
   time: number
 ): IndexIntoSamplesTable {
-  // TODO: This should really use bisect. samples.time is sorted.
-  for (let i: number = 0; i < samples.length; i++) {
-    if (samples.time[i] >= time) {
-      if (i === 0) {
-        return 0;
-      }
-      const distanceToThis = samples.time[i] - time;
-      const distanceToLast = time - samples.time[i - 1];
-      return distanceToThis < distanceToLast ? i : i - 1;
+  // samples.time is a sorted array of times, so we can bisect to find the latest sample
+  // before the provided time.
+  const bisectToLastSampleBeforeTime = (start, end) => {
+    if (start === end) {
+      // Bisect finished, sample found.
+      return start;
     }
-  }
-  return samples.length - 1;
+
+    const index = Math.ceil((end + start) / 2);
+    const sampleTime = samples.time[index];
+    if (sampleTime > time) {
+      // Bad sample, sample time superior to provided time
+      return bisectToLastSampleBeforeTime(start, index - 1);
+    } else {
+      // Correct sample, sample time inferior to provided time
+      return bisectToLastSampleBeforeTime(index, end);
+    }
+  };
+  return bisectToLastSampleBeforeTime(0, samples.length - 1);
```

Comment:
If you have a longer gap between two samples, this will always find the sample to the left of the gap, correct? It would be nice to find the sample whose center point (sample.time + interval/2) is closer to the mouse.

Reply:
Actually that's the root cause of the issue right? Based on the representation on the samples (see my comment in the original issue) this will lead to a user clicking on a bar, but selecting the bar to the right.

Reply:
Right. The sample bar rectangles have the left edge at sampleTime and the right edge at sampleTime + interval. The old code selects the sample based on the distance to sampleTime, i.e. based on the distance to the *left edge* of the rectangle.
I propose choosing the sample based on the distance to the *center* of the rectangle. The center of the rectangle is at sampleTime + (interval/2).

I think the alternative would be to draw these rectangles centered around the sampleTime, i.e. from sampleTime - interval / 2 to sampleTime + interval / 2. It's hard to say what's better; samples are taken at discrete points of time without any duration, and the only reason these bars have a width is that they'd be invisible otherwise, so there is no "true" answer to this question.

Reply:
Ah I read too fast sorry. I'll update my PR to do that.

One thing I was not sure about when changing this implementation was if there could be gaps in a samples collection. From your comment I assume gaps are possible, otherwise calculating the distance would always yield the same results as taking the closest previous time (I am assuming the interval is constant too ?).

If gaps are a thing that can happen between samples, for the purpose of clicking I guess it would be more sensible to check if the time is in [sample time, sample time + interval] rather than actually finding the shortest distance.

Reply:
Yes, there can be gaps between samples. Here's a profile that you can test your changes with: https://perfht.ml/2f9dx6M

The profiler tries to have a constant interval, but the reality can be a lot different, especially on Windows.